canal/sync.go

```diff
@@ -119,42 +111,25 @@ func (c *Canal) runSyncBinlog() error {
 				return errors.Trace(err)
 			}
 		case *replication.QueryEvent:
-			var (
-				mb    [][]byte
-				db    []byte
-				table []byte
-			)
-			regexps := []regexp.Regexp{*expCreateTable, *expAlterTable, *expRenameTable, *expDropTable, *expTruncateTable}
-			for _, reg := range regexps {
-				mb = reg.FindSubmatch(e.Query)
-				if len(mb) != 0 {
-					break
-				}
+			if e.GSet != nil {
+				c.master.UpdateGTIDSet(e.GSet)
 			}
-			mbLen := len(mb)
-			if mbLen == 0 {
+			stmts, _, err := c.parser.Parse(string(e.Query), "", "")
+			if err != nil {
+				log.Errorf("parse query(%s) err %v", e.Query, err)
 				continue
 			}
-
-			// the first last is table name, the second last is database name(if exists)
-			if len(mb[mbLen-2]) == 0 {
-				db = e.Schema
-			} else {
-				db = mb[mbLen-2]
-			}
-			table = mb[mbLen-1]
-
-			savePos = true
-			force = true
-			c.ClearTableCache(db, table)
-			log.Infof("table structure changed, clear table cache: %s.%s\n", db, table)
-			if err = c.eventHandler.OnTableChanged(string(db), string(table)); err != nil && errors.Cause(err) != schema.ErrTableNotExist {
-				return errors.Trace(err)
-			}
-
-			// Now we only handle Table Changed DDL, maybe we will support more later.
-			if err = c.eventHandler.OnDDL(pos, e); err != nil {
-				return errors.Trace(err)
+			for _, stmt := range stmts {
+				nodes := parseStmt(stmt)
+				if len(nodes) > 0 {
+					savePos = true
+					force = true
+				}
+				for _, node := range nodes {
+					if err = c.updateTable(node.db, node.table, e, pos, ev.Header.Timestamp); err != nil {
```

Comment:
It calls `updateTable` (`clear table cache` + `eventHandler.OnTableChanged` +  `c.eventHandler.OnDDL`) every table in ddl statement. It's ok for `clear table cache`  and `eventHandler.OnTableChanged`, but I think we should call `c.eventHandler.OnDDL` once for each ddl statement, right?  the reason is handler would get position in `OnDDL` and may save a position in `OnDDL`, so we should call `OnDDL` at end and only once